lib/jxl/modular/encoding/enc_ma.cc

```diff
@@ -37,33 +37,24 @@ const HWY_FULL(float) df;
 const HWY_FULL(int32_t) di;
 size_t Padded(size_t x) { return RoundUpTo(x, Lanes(df)); }
 
-float EstimateBits(const int32_t *counts, int32_t *rounded_counts,
-                   size_t num_symbols) {
-  // Try to approximate the effect of rounding up nonzero probabilities.
+// Compute entropy of the histogram, taking into account the minimum probability
+// for symbols with non-zero counts.
+float EstimateBits(const int32_t *counts, size_t num_symbols) {
   int32_t total = std::accumulate(counts, counts + num_symbols, 0);
-  const auto min = Set(di, (total + ANS_TAB_SIZE - 1) >> ANS_LOG_TAB_SIZE);
-  const auto zero_i = Zero(di);
-  for (size_t i = 0; i < num_symbols; i += Lanes(df)) {
-    auto counts_v = LoadU(di, &counts[i]);
-    counts_v = IfThenElse(Eq(counts_v, zero_i), zero_i,
-                          IfThenElse(Lt(counts_v, min), min, counts_v));
-    StoreU(counts_v, di, &rounded_counts[i]);
-  }
-  // Compute entropy of the "rounded" probabilities.
   const auto zero = Zero(df);
-  const size_t total_scalar =
-      std::accumulate(rounded_counts, rounded_counts + num_symbols, 0);
-  const auto inv_total = Set(df, 1.0f / total_scalar);
+  const auto minprob = Set(df, 1.0f / ANS_TAB_SIZE);
+  const auto inv_total = Set(df, 1.0f / total);
   auto bits_lanes = Zero(df);
-  auto total_v = Set(di, total_scalar);
+  auto total_v = Set(di, total);
   for (size_t i = 0; i < num_symbols; i += Lanes(df)) {
-    const auto counts_v = ConvertTo(df, LoadU(di, &counts[i]));
-    const auto round_counts_v = LoadU(di, &rounded_counts[i]);
-    const auto probs = Mul(ConvertTo(df, round_counts_v), inv_total);
-    const auto nbps = IfThenElse(Eq(round_counts_v, total_v), BitCast(di, zero),
-                                 BitCast(di, FastLog2f(df, probs)));
-    bits_lanes = Sub(bits_lanes, IfThenElse(Eq(counts_v, zero), zero,
-                                            Mul(counts_v, BitCast(df, nbps))));
+    const auto counts_iv = LoadU(di, &counts[i]);
+    const auto counts_fv = ConvertTo(df, counts_iv);
+    const auto probs = Mul(counts_fv, inv_total);
+    const auto mprobs = Max(probs, minprob);
+    const auto nbps = IfThenElse(Eq(counts_iv, total_v), BitCast(di, zero),
+                                 BitCast(di, FastLog2f(df, mprobs)));
+    bits_lanes = Sub(bits_lanes, IfThenElse(Eq(counts_fv, zero), zero,
```

Comment:
I think we don't need this "if 0" anymore - log(min_prob) is no longer NaN.

Reply:
Ah right, well spotted, that should make it a little bit faster again then..